README
NAME
    Mandel - Async model layer for MongoDB objects using Mango

VERSION
    0.28

SYNOPSIS
      # create your custom model class
      package MyModel;
      use Mojo::Base "Mandel";
      1;

      # create a document class
      package MyModel::Cat;
      use Mandel::Document;
      use Types::Standard 'Str';
      field name => ( isa => Str, builder => sub { "value" } );
      field 'type';
      belongs_to person => 'MyModel::Person';
      1;

      # create another document class
      package MyModel::Person;
      use Mandel::Document;
      use Types::Standard 'Int';
      field [qw( name )];
      field age => ( isa => Int );
      has_many cats => 'MyModel::Cat';
      has_one favorite_cat => 'MyModel::Cat';
      1;

      # use the model in your application
      package main;
      my $connection = MyModel->connect("mongodb://localhost/my_db");
      my $persons = $connection->collection('person');

      my $p1 = $persons->create({ name => 'Bruce', age => 30 });
      $p1->save(sub {
        my($p1, $err) = @_;
      });

      $persons->count(sub {
        my($persons, $err, $n_persons) = @_;
      });

      $persons->all(sub {
        my($persons, $err, $objs) = @_;
        for my $p (@$objs) {
          $p->age(25)->save(sub {});
        }
      });

      $persons->search({ name => 'Bruce' })->single(sub {
        my($persons, $err, $person) = @_;

        $person->cats(sub {
          my($person, $err, $cats) = @_;
          $_->remove(sub {}) for @$cats;
        });

        $person->remove(sub {
          my($person, $err) = @_;
        });
      });

DESCRIPTION
    THIS IS ALPHA SOFTWARE! THE API MAY BE CHANGED AT ANY TIME! PLEASE
    CONTACT ME IF YOU HAVE ANY COMMENTS OR FEEDBACK.

    Mandel is an async object-document-mapper. It allows you to work with
    your MongoDB documents in Perl as objects.

    This class binds it all together:

    *   Mandel::Model

        An object modelling a document.

    *   Mandel::Collection

        A collection of Mandel documents.

    *   Mandel::Document

        A single MongoDB document with logic.

ATTRIBUTES
    Mandel inherits all attributes from Mojo::Base and implements the
    following new ones.

  namespaces
    The namespaces which will be searched when looking for Types. By
    default, the (sub)class name of this module.

  model_class
    Returns Mandel::Model.

  storage
    An instance of Mango which acts as the database connection. If not
    provided.

METHODS
    Mandel inherits all methods from Mojo::Base and implements the following
    new ones.

  connect
      $self = $class->connect(@connect_args);
      $clone = $self->connect(@connect_args);

    @connect_args will be passed on to "new" in Mango, which again will be
    set as "storage".

    Calling this on an object will return a clone, but with a fresh
    "storage" object.

  all_document_names
      @names = $self->all_document_names;

    Returns a list of all the documents in the "namespaces".

  class_for
      $document_class = $self->class_for($name);

    Given a document name, find the related class name, ensure that it is
    loaded (or else die) and return it.

  collection
      $collection_obj = $self->collection($name);

    Returns a Mango::Collection object.

  model
      $model = $self->model($name);
      $self = $self->model($name => \%model_args);
      $self = $self->model($name => $model_obj);

    Define or returns a Mandel::Model object. Will die unless a model is
    registered by that name or "class_for" returns a class which has the
    "model()" method defined.

  initialize
      $self->initialize(@names, \%args);
      $self->initialize(\%args);

    Takes a list of document names. Calls the "initialize" in
    Mandel::Document method on any document given as input. @names default
    to "all_document_names" unless specified.

    %args defaults to empty hash ref, unless specified as input.

    The "initialize()" method will be called like this:

      $document_class->initialize($self, \%args);

SEE ALSO
    Mojolicious, Mango

    Still got MongoDB 2.4 on Ubuntu? Check out
    <http://docs.mongodb.org/manual/tutorial/install-mongodb-on-ubuntu/> to
    upgrade.

SOURCE REPOSITORY
    <http://github.com/jhthorsen/mandel>

AUTHORS
    Jan Henning Thorsen - "jhthorsen@cpan.org"

    Joel Berger - "joel.a.berger@gmail.com"

    Holger Rupprecht - "Holger.Rupprecht@Itelligence.de"

    Huo Linhe - "zitsen@cpan.org"

    This project is a fork of MangoModel
    <http://github.com/jberger/MangoModel>, created by Joel Berger,
    "joel.a.berger@gmail.com".

COPYRIGHT AND LICENSE
    Copyright (C) 2013 by Jan Henning Thorsen

    This library is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.



Changes
Revision history for perl distribution Mandel

0.28 2015-10-22T10:22:09+0200
 - Fix adding already saved object to belongsTo relationship #27
   Contributor: Huo Linhe
 - Fix adding already saved object to hasMany relationship #27
   Contributor: Huo Linhe

0.27 2015-04-15T08:14:26+0200
 - Deprecated plain "use Mandel;"

0.26 2015-04-14T10:53:52+0200
 - Add coercion support to fields #24
   Contributor: Huo Linhe

0.25 2015-03-01T17:56:17Z
 - Fixed Document->get and Document->contains

0.24 2015-03-01T17:52:44Z
 - Compatible with Mojolicious 6.0

0.23 2014-09-11T13:04:45Z
 - Fix pushing object to ListOf relationship
 - Change return value for BelongsTo get accessor: Now returns the saved
   object.
 - Change return value for HasOne get accessor: Now returns the saved
   object.
 - Mandel::Document will cache objects

0.22 2014-09-10T16:40:05Z
 - Fix missing dependencies
 - Run perltidy on the codebase

0.21 2014-09-10T15:50:50Z
 - Fix document extends Contributor: Alexey Neyasov

0.20 2014-05-08T19:26:11Z
 - Add "builder" support to field.
 - Remove "_build_data" from documentation.

0.19 2014-04-26T18:28:58Z
 - Fix DBRef in database: "$ref" did not have the right collection name

0.18 2014-04-22T07:44:14Z
 - Add EXPERIMENTAL ListOf relationship

0.17 2014-04-14T17:14:45Z
 - Improved DEBUG mode.

0.16 2014-03-22T13:48:14Z
 - Change use Mandel; will be deprecated
 - Fix annoying perl warning on "use Mandel"

0.15 2014-03-12T18:35:49Z
 - Add more blocking methods Contributors: Joel Berger and Alexey Neyasov
   https://github.com/jhthorsen/mandel/pull/9
   https://github.com/jhthorsen/mandel/pull/12

0.14 2014-03-05T06:57:04Z
 - Add Mandel::Document::TO_JSON() Contributor: Sebastian Riedel -
   https://github.com/jhthorsen/mandel/issues/10

0.13 2014-03-02T16:25:48Z
 - Add missing exception handling in HasOne relationship Contributor:
   Sebastian Riedel - https://github.com/jhthorsen/mandel/issues/7
 - Fix $col->remove() received $doc instead of $err in callback
 - Fix documentation error Contributor: Alexey Neyasov -
   https://github.com/jhthorsen/mandel/issues/8

0.12 2014-02-20T14:28:55Z
 - Add _build_data() for $document->data() default values
 - Improved documentation

0.11 2013-11-18T10:52:59Z
 - Change Mandel::Model->field(): %meta need to be a hash-ref
 - Change Mandel::Model->field(): Can retrive meta data for field
 - Add Mandel::Model::Field

0.10 2013-10-14T20:20:10Z
 - Change Mandel::Collection::patch() works on search() arguments
 - Add Mandel::Document->patch for updating partial documents

0.09 2013-10-13T18:29:40Z
 - Improved blocking support
 - Fix Document->remove() and ->save(): Forgot to pass on $err

0.08 2013-10-09T21:44:53Z
 - Add support for setting belongs_to accessor to oid
 - Fix Model->name() was not set when built from documents on disk

0.07 2013-10-06T14:46:55Z
 - Rename Model::add_field() to just field()
 - Rename Document::_raw() to ::data()
 - Relationships are now objects
 - Change related documents are stored using a dbref

0.06 2013-10-05T15:43:51Z
 - Add support for isa. This also will force numbers to be coerced from
   strings to actual numbers.
 - Add support for perl -Mandel -e'...'
 - Add support for blocking method call

0.05 2013-10-05T13:02:14Z
 - Change has_one/has_many documents point back to the "owner".
 - Add belongs_to relationship

0.04 2013-10-01T17:08:53Z
 - Add support for has_one() and has_many('/json/pointer')
 - Add support for MANDEL_CURSOR_DEBUG=1
 - Add MYMETA* to MANIFEST.SKIP
 - Add Model::new_collection()
 - Rename Model::collection() to Model::collection_name

0.03 2013-10-01T12:39:20Z
 - Fix default collection name on custom document base class
 - Add Mandel::Collection::patch()

0.02 2013-09-28T12:02:38Z
 - Add Mandel::Collection::save()
 - Add Mandel::Document::contains()
 - Add Mandel::connect(), remove ::uri()
 - Add Mandel::model_class()
 - Add Mandel::storage(), remove ::mango()
 - Add builders for attributes, which makes it easier to customize in
   subclass.
 - Add support for custom base class for documents

0.01 2013-09-27T18:55:40Z
 - Released alpha software
 - Add document model
 - Add document base class
 - Add collection managment
 - Add cursor
 - Add support for basic relationship: has_one and has_many



lib/Mandel.pm
package Mandel;

=head1 NAME

Mandel - Async model layer for MongoDB objects using Mango

=head1 VERSION

0.28

=head1 SYNOPSIS

  # create your custom model class
  package MyModel;
  use Mojo::Base "Mandel";
  1;

  # create a document class
  package MyModel::Cat;
  use Mandel::Document;
  use Types::Standard 'Str';
  field name => ( isa => Str, builder => sub { "value" } );
  field 'type';
  belongs_to person => 'MyModel::Person';
  1;

  # create another document class
  package MyModel::Person;
  use Mandel::Document;
  use Types::Standard 'Int';
  field [qw( name )];
  field age => ( isa => Int );
  has_many cats => 'MyModel::Cat';
  has_one favorite_cat => 'MyModel::Cat';
  1;

  # use the model in your application
  package main;
  my $connection = MyModel->connect("mongodb://localhost/my_db");
  my $persons = $connection->collection('person');

  my $p1 = $persons->create({ name => 'Bruce', age => 30 });
  $p1->save(sub {
    my($p1, $err) = @_;
  });

  $persons->count(sub {
    my($persons, $err, $n_persons) = @_;
  });

  $persons->all(sub {
    my($persons, $err, $objs) = @_;
    for my $p (@$objs) {
      $p->age(25)->save(sub {});
    }
  });

  $persons->search({ name => 'Bruce' })->single(sub {
    my($persons, $err, $person) = @_;

    $person->cats(sub {
      my($person, $err, $cats) = @_;
      $_->remove(sub {}) for @$cats;
    });

    $person->remove(sub {
      my($person, $err) = @_;
    });
  });

=head1 DESCRIPTION

THIS IS ALPHA SOFTWARE! THE API MAY BE CHANGED AT ANY TIME!
PLEASE CONTACT ME IF YOU HAVE ANY COMMENTS OR FEEDBACK.

L<Mandel> is an async object-document-mapper. It allows you to work with your
MongoDB documents in Perl as objects.

This class binds it all together:

=over 4

=item * L<Mandel::Model>

An object modelling a document.

=item * L<Mandel::Collection>

A collection of Mandel documents.

=item * L<Mandel::Document>

A single MongoDB document with logic.

=back

=cut

use Mojo::Base 'Mojo::Base';
use Mojo::Loader qw( find_modules load_class );
use Mojo::Util;
use Mandel::Collection;
use Mandel::Model;
use Mango;
use Carp 'confess';

our $VERSION = '0.28';

=head1 ATTRIBUTES

L<Mandel> inherits all attributes from L<Mojo::Base> and implements the
following new ones.

=head2 namespaces

The namespaces which will be searched when looking for Types. By default, the
(sub)class name of this module.

=head2 model_class

Returns L<Mandel::Model>.

=head2 storage

An instance of L<Mango> which acts as the database connection. If not
provided.

=cut

has model_class => sub { shift->_build_model_class };
has namespaces  => sub { shift->_build_namespaces };
has storage     => sub { shift->_build_storage };

sub _build_model_class {'Mandel::Model'}
sub _build_namespaces  { [ref $_[0]] }
sub _build_storage     { Mango->new('mongodb://localhost/mandeltest') }

=head1 METHODS

L<Mandel> inherits all methods from L<Mojo::Base> and implements the following
new ones.

=head2 connect

  $self = $class->connect(@connect_args);
  $clone = $self->connect(@connect_args);

C<@connect_args> will be passed on to L<Mango/new>, which again will be set
as L</storage>.

Calling this on an object will return a clone, but with a fresh L</storage>
object.

=cut

sub connect {
  my ($self, @args) = @_;
  my $storage = Mango->new(@args);

  if (my $class = ref $self) {
    $self = $class->new(%$self, storage => $storage);
  }
  else {
    $self = $self->new(storage => $storage);
  }

  $self;
}

=head2 all_document_names

  @names = $self->all_document_names;

Returns a list of all the documents in the L</namespaces>.

=cut

sub all_document_names {
  my $self = shift;
  my %names = map { $_, 1 } keys %{$self->{model} || {}};

  for my $ns (@{$self->namespaces}) {
    for my $name (find_modules $ns) {
      $name =~ s/^${ns}:://;
      $names{Mojo::Util::decamelize($name)} = 1;
    }
  }

  keys %names;
}

=head2 class_for

  $document_class = $self->class_for($name);

Given a document name, find the related class name, ensure that it is loaded
(or else die) and return it.

=cut

sub class_for {
  my ($self, $name) = @_;

  if (my $class = $self->{loaded}{$name}) {
    return $class;
  }

  for my $ns (@{$self->namespaces}) {
    my $class = $ns . '::' . Mojo::Util::camelize($name);
    my $e     = load_class $class;
    die $e if ref $e;
    next   if $e;
    return $self->{loaded}{$name} = $class;
  }

  confess "Could not find class for $name";
}

=head2 collection

  $collection_obj = $self->collection($name);

Returns a L<Mango::Collection> object.

=cut

sub collection {
  my ($self, $name) = @_;

  $self->model($name)->new_collection($self);
}

=head2 model

  $model = $self->model($name);
  $self = $self->model($name => \%model_args);
  $self = $self->model($name => $model_obj);

Define or returns a L<Mandel::Model> object. Will die unless a model is
registered by that name or L</class_for> returns a class which has the
C<model()> method defined.

=cut

sub model {
  my ($self, $name, $model) = @_;

  if ($model) {
    $model = $self->model_class->new($model) if ref $model eq 'HASH';
    $model->name($name);
    $self->{loaded}{$name} = $model->document_class;
    $self->{model}{$name}  = $model;
    return $self;
  }
  else {
    return $self->{model}{$name} ||= $self->class_for($name)->model;
  }
}

=head2 initialize

  $self->initialize(@names, \%args);
  $self->initialize(\%args);

Takes a list of document names. Calls the L<Mandel::Document/initialize> method
on any document given as input. C<@names> default to L</all_document_names>
unless specified.

C<%args> defaults to empty hash ref, unless specified as input.

The C<initialize()> method will be called like this:

  $document_class->initialize($self, \%args);

=cut

sub initialize {
  my $args      = ref $_[-1] eq 'HASH' ? pop : {};
  my $self      = shift;
  my @documents = @_ ? @_ : $self->all_document_names;

  for my $document (@documents) {
    my $class = $self->class_for($document);
    $class->initialize($self, $args);
  }
}

sub _storage_collection {
  $_[0]->storage->db->collection($_[1]);
}

=head1 SEE ALSO

L<Mojolicious>, L<Mango>

Still got MongoDB 2.4 on Ubuntu? Check out
L<http://docs.mongodb.org/manual/tutorial/install-mongodb-on-ubuntu/>
to upgrade.

=head1 SOURCE REPOSITORY

L<http://github.com/jhthorsen/mandel>

=head1 AUTHORS

Jan Henning Thorsen - C<jhthorsen@cpan.org>

Joel Berger - C<joel.a.berger@gmail.com>

Holger Rupprecht - C<Holger.Rupprecht@Itelligence.de>

Huo Linhe - C<zitsen@cpan.org>

This project is a fork of L<MangoModel|http://github.com/jberger/MangoModel>,
created by Joel Berger, C<joel.a.berger@gmail.com>.

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2013 by Jan Henning Thorsen

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=cut

1;
